### PR TITLE
Add ability to find terrain, furnitures, traps, monsters or npcs with EOC.

### PR DESCRIPTION
**Pull request by Ramza13**
#### Summary
None

#### Purpose of change
Request by @John-Candlebury and @Drew4484 
More json power

#### Describe the solution
Add new parameters to location_variable to allow finding terrain, furniture, traps, monsters or npcs.

#### Testing
Used this json
```
{
    "id": "EOC_SUCCESS",
    "type": "effect_on_condition",
    "effect": { "u_teleport": { "global_val": "test" } }
  },
  {
    "id": "EOC_FAIL",
    "type": "effect_on_condition",
    "effect": { "u_message": "Fail" }
  },
  {
    "type": "effect_on_condition",
    "id": "TEST_FURN",
    "effect": [
      {
        "u_location_variable": { "global_val": "test" },
        "target_params": { "om_terrain": "shelter_1", "search_range": 500, "z": 0 },
        "furniture": "f_bench",
        "target_max_radius": 30,
        "true_eocs": "EOC_SUCCESS",
        "false_eocs": "EOC_FAIL"
      }
    ]
  },
  {
    "type": "effect_on_condition",
    "id": "TEST_FURN_BLANK",
    "effect": [
      {
        "u_location_variable": { "global_val": "test" },
        "target_params": { "om_terrain": "shelter_1", "search_range": 500, "z": 0 },
        "furniture": "",
        "target_max_radius": 30,
        "true_eocs": "EOC_SUCCESS",
        "false_eocs": "EOC_FAIL"
      }
    ]
  },
  {
    "type": "effect_on_condition",
    "id": "TEST_TER",
    "effect": [
      {
        "u_location_variable": { "global_val": "test" },
        "target_params": { "om_terrain": "shelter_1", "search_range": 500, "z": 0 },
        "terrain": "t_pavement_y",
        "target_max_radius": 30,
        "min_radius": 1,
        "max_radius": 1,
        "true_eocs": "EOC_SUCCESS",
        "false_eocs": "EOC_FAIL"
      }
    ]
  },
  {
    "type": "effect_on_condition",
    "id": "TEST_TRAP",
    "effect": [
      {
        "u_location_variable": { "global_val": "test" },
        "target_params": { "om_terrain": "shelter_1", "search_range": 500, "z": 0 },
        "trap": "tr_beartrap",
        "target_max_radius": 30,
        "true_eocs": "EOC_SUCCESS",
        "false_eocs": "EOC_FAIL"
      }
    ]
  },
  {
    "type": "effect_on_condition",
    "id": "TEST_MONSTER",
    "effect": [
      {
        "u_location_variable": { "global_val": "test" },
        "target_params": { "om_terrain": "shelter_1", "search_range": 500, "z": 0 },
        "monster": "mon_zombie",
        "target_max_radius": 30,
        "x_adjust": 1,
        "true_eocs": "EOC_SUCCESS",
        "false_eocs": "EOC_FAIL"
      }
    ]
  },
  {
    "type": "effect_on_condition",
    "id": "TEST_MONSTER_BLANK",
    "effect": [
      {
        "u_location_variable": { "global_val": "test" },
        "target_params": { "om_terrain": "shelter_1", "search_range": 500, "z": 0 },
        "monster": "",
        "target_max_radius": 30,
        "x_adjust": 1,
        "true_eocs": "EOC_SUCCESS",
        "false_eocs": "EOC_FAIL"
      }
    ]
  },
  {
    "type": "effect_on_condition",
    "id": "TEST_NPC",
    "effect": [
      {
        "u_location_variable": { "global_val": "test" },
        "target_params": { "om_terrain": "shelter_1", "search_range": 500, "z": 0 },
        "npc": "NC_PORTAL_PERSON",
        "target_max_radius": 30,
        "x_adjust": 1,
        "true_eocs": "EOC_SUCCESS",
        "false_eocs": "EOC_FAIL"
      }
    ]
  },
  {
    "type": "effect_on_condition",
    "id": "TEST_NPC_BLANK",
    "effect": [
      {
        "u_location_variable": { "global_val": "test" },
        "target_params": { "om_terrain": "shelter_1", "search_range": 500, "z": 0 },
        "npc": "",
        "target_max_radius": 30,
        "x_adjust": 1,
        "true_eocs": "EOC_SUCCESS",
        "false_eocs": "EOC_FAIL"
      }
    ]
  },
  {
    "type": "effect_on_condition",
    "id": "TEST_ZONE",
    "effect": [
      {
        "u_location_variable": { "global_val": "test" },
        "target_params": { "om_terrain": "shelter_1", "search_range": 500, "z": 0 },
        "zone": "LOOT_PFOOD",
        "target_max_radius": 30,
        "x_adjust": 1,
        "true_eocs": "EOC_SUCCESS",
        "false_eocs": "EOC_FAIL"
      }
    ]
  },
  {
    "type": "effect_on_condition",
    "id": "TEST_FURN_CLOSE",
    "effect": [
      {
        "u_location_variable": { "global_val": "test" },
        "furniture": "f_bench",
        "target_min_radius": 3,
        "target_max_radius": 15,
        "true_eocs": "EOC_SUCCESS",
        "false_eocs": "EOC_FAIL"
      }
    ]
  },
  {
    "type": "effect_on_condition",
    "id": "TEST_TER_CLOSE",
    "effect": [
      {
        "u_location_variable": { "global_val": "test" },
        "terrain": "t_pavement_y",
        "target_min_radius": 3,
        "target_max_radius": 15,
        "min_radius": 1,
        "max_radius": 1,
        "true_eocs": "EOC_SUCCESS",
        "false_eocs": "EOC_FAIL"
      }
    ]
  },
  {
    "type": "effect_on_condition",
    "id": "TEST_TRAP_CLOSE",
    "effect": [
      {
        "u_location_variable": { "global_val": "test" },
        "trap": "tr_beartrap",
        "target_min_radius": 3,
        "target_max_radius": 15,
        "true_eocs": "EOC_SUCCESS",
        "false_eocs": "EOC_FAIL"
      }
    ]
  },
  {
    "type": "effect_on_condition",
    "id": "TEST_MONSTER_CLOSE",
    "effect": [
      {
        "u_location_variable": { "global_val": "test" },
        "monster": "mon_zombie",
        "target_min_radius": 3,
        "target_max_radius": 15,
        "x_adjust": 1,
        "true_eocs": "EOC_SUCCESS",
        "false_eocs": "EOC_FAIL"
      }
    ]
  },
  {
    "type": "effect_on_condition",
    "id": "TEST_NPC_CLOSE",
    "effect": [
      {
        "u_location_variable": { "global_val": "test" },
        "npc": "NC_PORTAL_PERSON",
        "target_min_radius": 3,
        "target_max_radius": 15,
        "x_adjust": 1,
        "true_eocs": "EOC_SUCCESS",
        "false_eocs": "EOC_FAIL"
      }
    ]
  },
  {
    "type": "effect_on_condition",
    "id": "TEST_ZONE_CLOSE",
    "effect": [
      {
        "u_location_variable": { "global_val": "test" },
        "zone": "LOOT_PFOOD",
        "target_min_radius": 3,
        "target_max_radius": 15,
        "x_adjust": 1,
        "true_eocs": "EOC_SUCCESS",
        "false_eocs": "EOC_FAIL"
      }
    ]
  }
```

### CODE REVIEW

**Comment by Termineitor244** [comment]
Awesome!! This allows/eases many ideas I had previously for scenarios, thank you so much!!

**Comment by Ramza13** [comment]
> Awesome!! This allows/eases many ideas I had previously for scenarios, thank you so much!!

Let me know if any more options would be helpful, there is always so much more this could do but I don't want to go too far beyond whats being used.

File: src/npctalk.cpp [hunk 1]
```diff
@@ -2694,38 +2694,139 @@ void talk_effect_fun_t::set_location_variable( const JsonObject &jo, const std::
...
         if( target_params.has_value() ) {
             const tripoint_abs_omt omt_pos = mission_util::get_om_terrain_pos( target_params.value() );
             target_pos = tripoint( project_to<coords::ms>( omt_pos ).x(), project_to<coords::ms>( omt_pos ).y(),
                                    project_to<coords::ms>( omt_pos ).z() );
-        } else if( max_radius > 0 ) {
+        }
+        const tripoint_abs_ms abs_ms( target_pos );
+        bool shifted = false;
+        if( !here.inbounds( abs_ms ) ) {
+            g->place_player_overmap( project_to<coords::omt>( abs_ms ), false );
```
Comment (andrei8l) [inline src/npctalk.cpp:2751]: Load a (tiny)map and check there please. Don't move the avatar around as it causes no end of trouble (ex: #60096 and the myriad of issues in the TCL).

You might have to manually spawn and unload monsters though.

**Comment by Termineitor244** [comment]
> > Awesome!! This allows/eases many ideas I had previously for scenarios, thank you so much!!
> 
> Let me know if any more options would be helpful, there is always so much more this could do but I don't want to go too far beyond whats being used.

The only thing I can think of that could be of use would be to be able to find zones too (like the ones the player can designate for sorting loot), I have an idea for a mod incorporating aspects of the defense mode for normal playthroughs (so suffering from periodic/random attacks from monsters to your static base without using the horde mechanic), and the player being able to designate a zone for the game to recognize as "closed space/no spawn zone" would be useful so monsters don't teleport inside the base of the player by mistake, an scenario I had in mind since several months ago would be equally easier to create with a custom zone for a safe haven base that does not spawn enemies with the EOC, but in that case it can be done too with the terrain/location id that you already present here, if only with certain limitations when doing the mapgen of the location. 

In short: Zones, finding certain types of zones would be perfect.

**Comment by Ramza13** [comment]
> > > Awesome!! This allows/eases many ideas I had previously for scenarios, thank you so much!!
> > 
> > 
> > Let me know if any more options would be helpful, there is always so much more this could do but I don't want to go too far beyond whats being used.
> 
> The only thing I can think of that could be of use would be to be able to find zones too (like the ones the player can designate for sorting loot), I have an idea for a mod incorporating aspects of the defense mode for normal playthroughs (so suffering from periodic/random attacks from monsters to your static base without using the horde mechanic), and the player being able to designate a zone for the game to recognize as "closed space/no spawn zone" would be useful so monsters don't teleport inside the base of the player by mistake, an scenario I had in mind since several months ago would be equally easier to create with a custom zone for a safe haven base that does not spawn enemies with the EOC, but in that case it can be done too with the terrain/location id that you already present here, if only with certain limitations when doing the mapgen of the location.
> 
> In short: Zones, finding certain types of zones would be perfect.

Ok added zones to the options.
